src/pd_gmres.m

```diff
@@ -0,0 +1,282 @@
+end
+
+% Check whether b has the same number of rows as b
+if rowsb ~= n
+    error("Dimension mismatch between matrix A and vector b.")
+end
+
+clear rowsb colsb;
+
+% ----> Default values and sanity checks for parameter m0
+
+% When the restart parameter is not specified explicitly as an input
+% argument, pd_gmres() behaves identically to the unrestarted gmres()
+% with a fixed number of iterations given by min(n, 10).
+%
+% There are three possibilities when this might happen: 
+%   (1) If only two arguments are given i.e., A and b.
+%   (2) If m0 equals the dimension of A i.e., m0 = n.
+%   (3) If an empty matrix is given as restart parameter, i.e., m0 = [].
+if (nargin < 3) || isempty(m0) || (m0 == n)
+    restarted = false;  % use unrestarted pd_gmres()
+else
+    restarted = true;  % use restarted pd_gmres()
+end
+
+if restarted
+    if m0 <= 0
+        error("Restart parameter m0 must be a positive integer.")
+    elseif m0 > n
+        error("Restart parameter m0 cannot be greater than n.")
+    end
+end
+
+% ----> Default value and sanity checks for tol
+if (nargin < 4) || isempty(tol)
+    tol = 1e-6;
+end
+
+if tol < eps
+    warning("Tolerance is too small and it will be changed to eps.")
+    tol = eps;
+elseif tol >= 1
+    warning("Tolerance is too large and it will be changed to 1-eps.")
+    tol = 1 - eps;
+end
+
+% ----> Default value for maxit
+if (nargin < 5) || isempty(maxit)
+    if restarted
+        % If the restarted version of pd_gmres() version must be used
+        % but maxit is not given, we take the min between n/m0 and 10.
+        maxit = min(ceil(n/m0), 10);
+    else
+        % If the unrestarted version must ...
+        maxit = min(n, 10);
+    end
+end
+    
+% ----> Default value and sanity checks for initial guess x0
+if (nargin < 6) || isempty(x0) 
+    x0 = zeros(n, 1);
+end
+
+% Check whether x0 is a column vector
+[rowsx0, colsx0] = size(x0);          
+if colsx0 ~= 1
+    error("Initial guess x0 is not a column vector.")
+end
+
+% Check whether x0 has the right dimension
+if rowsx0 ~= n
+    error("Initial guess x0 does not have the correct dimension.")
+end
+
+clear rowsx0 colsx0;
+
+% ----> Default value for propotional parameter alphaP
+if (nargin < 7) || isempty(alphaP)
+    alphaP = -3;
+end
+
+% ----> Default value for proportional parameter alphaD
+if (nargin < 8) || isempty(alphaD)
+    alphaD = 5;
+end
+
+% ---> Algorithm starts here 
+
+% Unrestarted version of the PD-GMRES
+% This block calls the bult-in gmres() function but outputs slightly
+% different variables. In particular, the resvec vector only shows the
+% initial and the last residual (contrary to gmres() that stores all the
+% residuals).
+if ~restarted
+    tic();
+    [x, flag, relres, iter, resvec] = gmres(A, b, [], tol, maxit, [], [], x0);
+    resvec = [resvec(1); resvec(end)];
+    time = toc();
+    return
+end
+
+% Restarted version of PD-GMRES
+
+% Algorithm setup 
+mmin = 1;
+mmax = n-1;
+mstep =1 ;

```

Comment:
This value should be discussed, in Cuevas' article it was considered mstep=3, but because the matrices were of large dimensions.

Reply:
After our conversation, this will be handled by #22.